demos/device_defender_for_aws/metrics_collector/lwip/metrics_collector.c

```diff
@@ -0,0 +1,254 @@
+            ( pOutNumTcpOpenPorts == NULL ) )
+        {
+            IotLogError( "Invalid parameters. pOutTcpPortsArray: 0x%08x,"
+                         "portsArrayLength: %u, pOutNumTcpOpenPorts: 0x%08x.",
+                         pOutTcpPortsArray,
+                         portsArrayLength,
+                         pOutNumTcpOpenPorts );
+            status = MetricsCollectorBadParameter;
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            LOCK_TCPIP_CORE();
+
+            while( pCurrPcb )
+            {
+                /*  Write the ports into pOutTcpPortsArray, if not NULL */
+                if( ( pOutTcpPortsArray != NULL ) && ( pcbCnt < portsArrayLength ) )
+                {
+                    pOutTcpPortsArray[ pcbCnt ] = pCurrPcb->local_port;
+                }
+
+                ++pcbCnt;
+                pCurrPcb = pCurrPcb->next;
+            }
+
+            UNLOCK_TCPIP_CORE();
+        }
+
+        if( ( pOutTcpPortsArray != NULL ) && ( pcbCnt > portsArrayLength ) )
+        {
+            IotLogWarn( "The portsArrayLength is not long engough to store all the listening ports" );
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            *pOutNumTcpOpenPorts = pcbCnt;
+        }
+
+        return status;
+    }
+
+
+    MetricsCollectorStatus_t GetOpenUdpPorts( uint16_t * pOutUdpPortsArray,
+                                              uint32_t portsArrayLength,
+                                              uint32_t * pOutNumUdpOpenPorts )
+    {
+        MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+        struct udp_pcb * pCurrPcb = udp_pcbs;
+        uint16_t pcbCnt = 0;
+
+        if( ( ( pOutUdpPortsArray != NULL ) && ( portsArrayLength == 0 ) ) ||
+            ( pOutNumUdpOpenPorts == NULL ) )
+        {
+            IotLogError( "Invalid parameters. pOutUdpPortsArray: 0x%08x,"
+                         "portsArrayLength: %u, pOutNumUdpOpenPorts: 0x%08x.",
+                         pOutUdpPortsArray,
+                         portsArrayLength,
+                         pOutNumUdpOpenPorts );
+            status = MetricsCollectorBadParameter;
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            LOCK_TCPIP_CORE();
+
+            while( pCurrPcb )
+            {
+                /*  Write the ports into pOutUdpPortsArray, if not NULL */
+                if( ( pOutUdpPortsArray != NULL ) && ( pcbCnt < portsArrayLength ) )
+                {
+                    pOutUdpPortsArray[ pcbCnt ] = pCurrPcb->local_port;
+                }
+
+                ++pcbCnt;
+                pCurrPcb = pCurrPcb->next;
+            }
+
+            UNLOCK_TCPIP_CORE();
+        }
+
+        if( ( pOutUdpPortsArray != NULL ) && ( pcbCnt > portsArrayLength ) )
+        {
+            IotLogWarn( "The portsArrayLength is not long engough to store all the listening ports" );
```

Comment:
enough and please use LogWarn()

Reply:
Sorry forgot update the logs of metrics collector...

Reply:
done by commit dbfa0fc8c562ae295a49aa8905ac5c92457a91b0